### PR TITLE
perf(autoware_cuda_pointcloud_preprocessor): process points in ring order instead of an organized layout

### PR DESCRIPTION
**Pull request by mojomex**
## Description

Fixes the `cuda_pointcloud_preprocessor` performance regression from #13134/#13135 by dropping the organized `rings x points-per-ring` layout altogether.

**Regression** (RTX 4000 SFF Ada, eight preprocessor nodes at 10 Hz sharing the GPU with PTv3, whole-stack replay, Nsight Systems, per-invocation attribution by correlation id):
- the layout is fixed at `max_ring_count x max_points_per_ring` (921,600 slots for 128 x 7200), so every fill, kernel, the 4-pass segmented sort, the scan and the reductions run over the configured maximum, not the frame;
- three fills clear ~30 MB point buffers the next kernel overwrites anyway (`fillKernel`: 13 launches, 3.1 ms per invocation).

**Change**
- The organized layout only existed to give the ring outlier filter its ring neighbours; `organizeKernel` placed points with atomics (scrambling the order within a ring) and the segmented sort restored it. Both are replaced by one stable radix sort of the point indices by ring key (`cub::DeviceRadixSort::SortPairs`, 16-bit keys): rings contiguous, input (scan) order within a ring, a dense list of exactly the frame's points.
- `transformPointsKernel` gathers through the sorted indices: no organized copy, no gather kernel, no validity mask, no sentinel fill. The ring outlier filter walks the dense list and clamps its window to same-channel neighbours.
- Every kernel, the scan and the reductions cover `num_raw_points_`, known on the host before the first launch; buffers stay preallocated at `max_input_point_count`. No runtime allocation, no added synchronization (the one after the input copy is removed; it guarded nothing).
- Rings have no capacity any more: `max_ring_count` and `max_points_per_ring` are removed, no point is dropped for its ring index or ring size, the `ring_overflow` diagnostic goes; input truncation to `max_input_point_count` and its diagnostic stay. New diagnostic: the frame's largest ring (`Maximum points per ring`).
- Output order is unchanged (ring-major, scan order within a ring).

**Result** (three replicates each; ranges are replicate medians)

| per preprocessor invocation / system | before #13134 | #13134/#13135 | this PR |
|---|---|---|---|
| preprocessor GPU active | 0.96–0.99 ms | 4.86–5.43 ms | **0.31–0.32 ms** |
| preprocessor kernel time (sum) | 1.47–1.49 ms | 5.60–5.91 ms | **0.44–0.48 ms** |
| preprocessor node latency | 10.5 ms | 9.0–9.4 ms | **1.8–2.1 ms** |
| PTv3 latency (median) | 88.1–88.8 ms | 109–113 ms | **74–80 ms** |
| PTv3 dropped frames (vs 10 Hz) | 2.9–4.8 % | 18.6–18.9 % | **0.2–0.8 %** |
| GPU busy (trace) | 71.4–72.0 % | 84.2–85.5 % | **65.9–67.7 %** |

## Related links

**Parent Issue:**

- #13134, #13135 (the regression); #13137 (tests this builds on)

## How was this PR tested?

- `colcon test --packages-select autoware_cuda_pointcloud_preprocessor` in the benchmark image on the RTX 4000 SFF Ada: 14 tests, 0 failures, 0 skipped.
- New test `OutputDoesNotDependOnEarlierLargerFrames`: frames of growing, shrinking, truncated and normal size processed in sequence with non-empty twist/IMU queues, each compared byte for byte (and by stats) with a fresh preprocessor's output for the same input.
- #13137's per-ring-capacity tests replaced by `KeepsEveryPointOfEveryRing` and `OrdersOutputByRingThenInputOrder`.
- Output equivalence against `main` on 40 synthetic 40-ring frames x 4 filter configurations (2D/3D undistortion, crop on/off, ring filter on/off): output clouds byte-identical wherever no per-ring drop applied on `main`.
- Real inputs: the eight preprocessor outputs of a 60 s replay of the benchmark bag (0.1x, deterministic) recorded on the stack before #13134 and on this PR: 4,678 of 4,790 frames byte-identical, 109 frames with a few points shifted <= 2 mm (twist-sample timing), 3 first frames where the old stack lost 4-55 % of its points and this PR keeps them (details and video in the comment below).
- Whole-stack benchmark above: all eight nodes publish at 10 Hz in every replicate, no crashes.

## Notes for reviewers

- Per-ring drops no longer exist; over-capacity handling is only the existing truncation to `max_input_point_count`.
- With no crop box configured, `num_crop_box_passed_points` now equals the point count (it used to count layout slots).

#### Additions and removals

| Change type | Parameter Name        | Type      | Default Value | Description |
|:------------|:----------------------|:----------|:--------------|:------------|
| Removed     | `max_ring_count`      | `integer` | `128`         | Rings have no capacity any more; the value is ignored if still passed. |
| Removed     | `max_points_per_ring` | `integer` | `7200`        | Same. |

Diagnostics: the key `Ring organization overflow` is removed from the input-bounds status; `Maximum points per ring` is added.

## Effects on system behavior

- The preprocessor's GPU time follows the frame's point count; the models sharing the GPU regain the time (numbers above).
- Points on rings beyond the previous `max_ring_count`, or beyond `max_points_per_ring` in one ring, are now kept instead of dropped.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: sensing/autoware_cuda_pointcloud_preprocessor/src/cuda_pointcloud_preprocessor/organize_kernels.cu [hunk 1]
```diff
@@ -44,7 +45,7 @@ __global__ void organizeKernel(
   int next_offset = atomicAdd(&ring_indexes[ring], 1);
 
   if (next_offset >= initial_max_points_per_ring) {
-    atomicOr(output_points_per_ring_overflowed, 1);
+    atomicMax(output_points_per_ring_overflowed, next_offset + 1);
```
Comment (amadeuszsz) [inline sensing/autoware_cuda_pointcloud_preprocessor/src/cuda_pointcloud_preprocessor/organize_kernels.cu]: nit:
This is not flag anymore, right? So maybe `output_overflow_points_per_ring`.

File: sensing/autoware_cuda_pointcloud_preprocessor/test/test_cuda_pointcloud_preprocessor.cpp [hunk 2]
```diff
@@ -354,6 +401,73 @@ TEST(QueueBounds, RejectsAlreadyOverCapacityInternalQueue)
...
+      EXPECT_LE(from_fresh.width, static_cast<std::uint32_t>(max_points_per_ring));
+      continue;
+    }
+    EXPECT_EQ(from_sequence.width, from_fresh.width) << "frame " << i;
+    EXPECT_EQ(from_sequence.data, from_fresh.data) << "frame " << i;
+    EXPECT_EQ(
+      from_sequence.stats.num_crop_box_passed_points, from_fresh.stats.num_crop_box_passed_points)
+      << "frame " << i;
+    EXPECT_EQ(from_sequence.stats.num_nan_points, from_fresh.stats.num_nan_points) << "frame " << i;
+    EXPECT_EQ(from_sequence.stats.mismatch_count, from_fresh.stats.mismatch_count) << "frame " << i;
```
Comment (amadeuszsz) [inline sensing/autoware_cuda_pointcloud_preprocessor/test/test_cuda_pointcloud_preprocessor.cpp:416]: nit:

Might be worth to pass non-empty queues in 186:189, otherwise this check is pointless.

File: sensing/autoware_cuda_pointcloud_preprocessor/test/test_cuda_pointcloud_preprocessor.cpp [hunk 3]
```diff
@@ -154,6 +155,52 @@ std::shared_ptr<const sensor_msgs::msg::Imu> make_imu_ptr(const std::uint32_t na
...
+{
+  const auto output_cloud = preprocessor.process(
+    input_cloud, make_identity_transform(),
+    std::deque<geometry_msgs::msg::TwistWithCovarianceStamped>{},
+    std::deque<geometry_msgs::msg::Vector3Stamped>{}, 0U);
+  ProcessedCloud processed;
+  processed.width = output_cloud->width;
+  processed.data.resize(output_cloud->row_step);
+  if (!processed.data.empty()) {
+    cudaMemcpy(
```
Comment (amadeuszsz) [inline sensing/autoware_cuda_pointcloud_preprocessor/test/test_cuda_pointcloud_preprocessor.cpp]: Shouldn't be wrapped with `CHECK_CUDA_ERROR`?

File: sensing/autoware_cuda_pointcloud_preprocessor/src/cuda_pointcloud_preprocessor/cuda_pointcloud_preprocessor.cu [hunk 4]
```diff
@@ -518,6 +527,20 @@ std::unique_ptr<cuda_blackboard::CudaPointCloud2> CudaPointcloudPreprocessor::pr
...
   CHECK_CUDA_ERROR(cudaStreamSynchronize(stream_));
+
+  // A ring held more points than the organized extent covers: widen it (buffers already hold
+  // the capacity, only the layout changes) and redo this frame. The extent grows monotonically
+  // up to the capacity, so this happens a handful of times after startup and never in steady
+  // state; at the capacity the overflow is reported as ring_overflow instead.
+  if (points_per_ring_overflowed != 0 && max_points_per_ring_ < capacity_.max_points_per_ring) {
+    const int rounded = (points_per_ring_overflowed + points_per_ring_granularity - 1) /
+                        points_per_ring_granularity * points_per_ring_granularity;
+    setMaxPointsPerRing(std::min(capacity_.max_points_per_ring, rounded));
```
Comment (amadeuszsz) [inline sensing/autoware_cuda_pointcloud_preprocessor/src/cuda_pointcloud_preprocessor/cuda_pointcloud_preprocessor.cu]: This never shrinks and affects the performance. Mechanics for lowering max point per ring seems to complex, can cause more troubles, forget about it. Consider adding the current value of max point per ring to diagnostics output.

**Review by amadeuszsz — CHANGES_REQUESTED** [review]

File: sensing/autoware_cuda_pointcloud_preprocessor/src/cuda_pointcloud_preprocessor/organize_kernels.cu (hunk 1 above)
Comment (mojomex) [reply sensing/autoware_cuda_pointcloud_preprocessor/src/cuda_pointcloud_preprocessor/organize_kernels.cu]: Applied in spirit, the output itself is gone: the rewrite (161319c3bc) drops the organized layout, so there is no per-ring overflow output any more. Rings have no capacity: points are stably sorted by ring key (`cub::DeviceRadixSort::SortPairs`) and processed as a dense list of exactly the frame's points, and `max_ring_count` / `max_points_per_ring` are removed.

File: sensing/autoware_cuda_pointcloud_preprocessor/test/test_cuda_pointcloud_preprocessor.cpp (hunk 2 above)
Comment (mojomex) [reply sensing/autoware_cuda_pointcloud_preprocessor/test/test_cuda_pointcloud_preprocessor.cpp:416]: Done: `OutputDoesNotDependOnEarlierLargerFrames` now passes two twist and two IMU samples stamped after the cloud, so the 2D undistortion runs for every frame of the sequence (and `mismatch_count` is compared as well). `process_cloud` takes the queues as arguments.

File: sensing/autoware_cuda_pointcloud_preprocessor/test/test_cuda_pointcloud_preprocessor.cpp (hunk 3 above)
Comment (mojomex) [reply sensing/autoware_cuda_pointcloud_preprocessor/test/test_cuda_pointcloud_preprocessor.cpp]: Done: the `cudaMemcpy` in the test helper is wrapped in `CHECK_CUDA_ERROR` (`autoware/cuda_utils/cuda_check_error.hpp`).

File: sensing/autoware_cuda_pointcloud_preprocessor/src/cuda_pointcloud_preprocessor/cuda_pointcloud_preprocessor.cu (hunk 4 above)
Comment (mojomex) [reply sensing/autoware_cuda_pointcloud_preprocessor/src/cuda_pointcloud_preprocessor/cuda_pointcloud_preprocessor.cu]: Agreed, and it is gone: the rewrite has no running maximum and no per-ring extent at all. Every kernel, the sort, the scan and the reductions cover exactly the frame's point count (known on the host before the first launch), so nothing sticky affects later frames. For the diagnostics part of your comment, the frame's largest ring is now reported: `ProcessingStats::max_points_per_ring` (a tiny kernel over the sorted ring keys, read back with the existing end-of-frame copy) shows up as "Maximum points per ring" in the input-bounds diagnostic.

**Comment by mojomex** [comment]
### Output equivalence on real inputs, and a side-by-side video

**Setup.** Replays of the 60 s benchmark bag (x2 vehicle, eight Hesai lidars) on the stack from right before #13134/#13135 (upstream `f93d3eb885`) and on this PR's head; `ros2 bag record` of the eight `/sensing/lidar/<pos>/pointcloud_before_sync` outputs inside the pilot container; frames paired by `header.stamp` and compared byte for byte (x y z intensity return_type channel, in order).

**Deterministic replay (`ros2 bag play` at 0.1x, so twist/IMU/cloud arrival order does not depend on scheduling), all 4,790 frames paired:**

| class | frames | what it is |
|---|---|---|
| identical | **4,678** (97.7 %) | same width, same bytes, same order |
| values-differ | 109 | same width/channels/order; a few points (median 2 % of the frame) shifted by ≤ 2 mm (p50 0.1 mm) — residual twist-sample timing, see below |
| other | 3 | the **first frame** of each 128-channel lidar: the pre-#13134 stack publishes it with 4 % (front_upper: 31,477 vs 32,916 points) to 55 % (left_upper 40,840 vs 91,050; right_upper 40,910 vs 90,852) of its points missing, this PR with all of them. Reproduced in every "before" replay (1x, 1x again, 0.1x); from the second frame on both stacks agree. Pre-existing start-up loss of the adaptive layout, removed by this PR. |
| after-superset (per-ring drops undone) | 0 | no ring exceeded 7,200 points or ring index 128 on this bag |

**Video** (`left_upper`, frames paired by stamp): 


https://github.com/user-attachments/assets/1e3770f1-b109-4c04-9343-6ae581e82ad6

File: sensing/autoware_cuda_pointcloud_preprocessor/src/cuda_pointcloud_preprocessor/cuda_pointcloud_preprocessor.cu
Comment (amadeuszsz) [inline sensing/autoware_cuda_pointcloud_preprocessor/src/cuda_pointcloud_preprocessor/cuda_pointcloud_preprocessor.cu:285]: Please check if `stats_` now needs reset due to this early return. Otherwise you will log stale maximum points per ring.

File: sensing/autoware_cuda_pointcloud_preprocessor/test/test_cuda_pointcloud_preprocessor.cpp
Comment (amadeuszsz) [inline sensing/autoware_cuda_pointcloud_preprocessor/test/test_cuda_pointcloud_preprocessor.cpp:1]: nit:
Consider unit tests with more edge cases regarding points per ring, eg.:
```
{100, 0, 0, 1}

{1, 0, 0, 100}

{1, 1, 1, 1}

{0, 0, 0, 1}

{100, 1, 100}
```
Even though your code works, this is for sake of potential future changes.

**Review by amadeuszsz — APPROVED** [review]
LGTM overall
